larpandoracontent/LArThreeDReco/LArHitCreation/RANSACMethod.cc

```diff
@@ -0,0 +1,476 @@
+    const int secondaryTotal(estimator.GetSecondBestInliers().size() + secondModelCount + secondaryFavoured);
+
+    const RANSACHitVector bestResult(primaryTotal > secondaryTotal ? primaryResult : secondaryResult);
+
+    for (auto hit : bestResult)
+        protoHitVector.push_back(hit.GetProtoHit());
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void RANSACMethodTool::GetCandidatePoints(RANSACHitVector &allHits, ParameterVector &candidatePoints)
+{
+    if (allHits.size() < 1000)
+    {
+        for (auto hit : allHits)
+            candidatePoints.push_back(std::make_shared<RANSACHit>(hit));
+
+        return;
+    }
+
+    std::mt19937 eng(allHits.size());
+    const unsigned int hitsToUse((allHits.size() - 1) * 0.40);
+
+    // INFO: Fisher-Yates shuffle to get N unique random elements.
+    for (unsigned int i = 0; i <= hitsToUse; ++i)
+    {
+        int j(GetIntsInRange(i, allHits.size() - 1, eng));
+        RANSACHit pointI = allHits[i];
+        allHits[i] = allHits[j];
+        allHits[j] = pointI;
+
+        candidatePoints.push_back(std::make_shared<RANSACHit>(allHits[i]));
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+int RANSACMethodTool::RunOverRANSACOutput(RANSAC<PlaneModel, 3> &ransac, RANSACResult run, RANSACHitVector &hitsToUse, RANSACHitVector &finalHits)
+{
+    std::map<const CaloHit*, RANSACHit> inlyingHitMap;
+
+    if (run == RANSACResult::Second && !ransac.GetSecondBestModel())
+        return 0;
+
+    const ParameterVector currentInliers = run == RANSACResult::Best ? ransac.GetBestInliers() : ransac.GetSecondBestInliers();
+
+    // ATTN: The second model can technically be NULL, so set to first model in that case, as that won't cause any issues.
+    //       This is because its possible for the second model to not be populated (if say the first model fits everything),
+    //       but this method will not run at all if no models were populated.
+    const auto bestModel = ransac.GetBestModel().get();
+    const auto secondModel = ransac.GetSecondBestModel() ? ransac.GetSecondBestModel().get() : bestModel;
+    const auto currentModel = run == RANSACResult::Best ? bestModel : secondModel;
+    const auto otherModel = run == RANSACResult::Best ? secondModel : bestModel;
+
+    RANSACHitVector sortedHits;
+    std::list<RANSACHit> nextHits;
+
+    for (auto inlier : currentInliers)
+    {
+        const auto hit = std::dynamic_pointer_cast<RANSACHit>(inlier);
+
+        if (hit == nullptr)
+            throw std::runtime_error("Inlying hit was not of type RANSACHit");
+
+        const bool addedHit(RANSACMethodTool::AddToHitMap(*hit, inlyingHitMap));
+
+        if (addedHit)
+            sortedHits.push_back(*hit);
```

Comment:
I'm having a hard time understanding the relationship between some of the hits here. If this protohit is deemed a better candidate than the existing protohit for some parent calo hit, it seems that this and any preceding protohits all get add to the list of sorted hits, which suggests we, in principle at least, could end up with multiple protohits representing the same parent calo hit in one set of sorted hits? Is that right? The final hits list seems to avoid this problem. So what'a the goal of this sorted list?

Reply:
I second that! I think it would really help to have a few lines of comment to help clarify the flow of this function!